lib/modules/manager/gomod/artifacts.spec.ts

```diff
@@ -1289,4 +1295,271 @@ describe('modules/manager/gomod/artifacts', () => {
+      { file: { type: 'addition', path: 'main.go', contents: 'New main.go' } },
+      { file: { type: 'addition', path: 'go.mod', contents: 'New go.mod' } },
+    ]);
+    expect(execSnapshots).toMatchObject([
+      {
+        cmd: 'go get -d -t ./...',
+        options: { cwd: '/tmp/github/some/repo' },
+      },
+      {
+        cmd: 'go install github.com/marwan-at-work/mod/cmd/mod@latest',
+        options: { cwd: '/tmp/github/some/repo' },
+      },
+      {
+        cmd: 'mod upgrade --mod-name=github.com/google/go-github/v24 -t=28',
+        options: { cwd: '/tmp/github/some/repo' },
+      },
+      {
+        cmd: 'go mod tidy',
+        options: { cwd: '/tmp/github/some/repo' },
+      },
+      {
+        cmd: 'go mod tidy',
+        options: { cwd: '/tmp/github/some/repo' },
+      },
+    ]);
+  });
+
+  it('go.mod file contains go version', async () => {
+    GlobalConfig.set({ ...adminConfig, binarySource: 'docker' });
+    fs.readLocalFile.mockResolvedValueOnce('Current go.sum');
+    fs.readLocalFile.mockResolvedValueOnce(null); // vendor modules filename
+    fs.readLocalFile.mockResolvedValueOnce('someText\n\ngo 1.17\n\n');
+    const execSnapshots = mockExecAll();
+    git.getRepoStatus.mockResolvedValueOnce({
+      modified: ['go.sum', 'main.go'],
+    } as StatusResult);
+    fs.readLocalFile
+      .mockResolvedValueOnce('New go.sum')
+      .mockResolvedValueOnce('New main.go')
+      .mockResolvedValueOnce('New go.mod');
+    const exec = jest.spyOn(execUtl, 'exec');
+    const res = await gomod.updateArtifacts({
+      packageFileName: 'go.mod',
+      updatedDeps: [{ depName: 'github.com/google/go-github/v24' }],
+      newPackageFileContent: gomod1,
+      config: {
+        ...config,
+        updateType: 'major',
+        newMajor: 28,
+        postUpdateOptions: ['gomodUpdateImportPaths'],
+        constraints: {
+          gomodMod: 'a.b.c',
+        },
+      },
+    });
+
+    expect(res).toEqual([
+      { file: { type: 'addition', path: 'go.sum', contents: 'New go.sum' } },
+      { file: { type: 'addition', path: 'main.go', contents: 'New main.go' } },
+      { file: { type: 'addition', path: 'go.mod', contents: 'New go.mod' } },
+    ]);
+    expect(execSnapshots).toMatchObject([
+      {
+        cmd: 'docker pull renovate/go:latest',
+        options: { encoding: 'utf-8' },
+      },
+      {
+        cmd: 'docker ps --filter name=renovate_go -aq',
+        options: { encoding: 'utf-8' },
+      },
+      {
+        cmd:
+          'docker run --rm --name=renovate_go --label=renovate_child -v "/tmp/github/some/repo":"' +
+          '/tmp/github/some/repo" -v "/tmp/renovate/cache":"/tmp/renovate/cache" -e GOPROXY -e GOPRIVATE -e ' +
+          'GONOPROXY -e GONOSUMDB -e GOINSECURE -e GOFLAGS -e CGO_ENABLED -e BUILDPACK_CACHE_DIR -w ' +
+          '"/tmp/github/some/repo" renovate/go:latest bash -l -c "go get -d -t ./... && ' +
```

Comment:
mock datasource return value, so it's using a real version instead of latest.

Reply:
All the old tests are using `latest` in the command.

do you have any suggestions on how to do the change you requested?

Reply:
checkout other managers with artifacts handling, like python.

Reply:
@viceice I tried to use mocked datasource by adding this to my test:
```
    datasource.getPkgReleases.mockResolvedValueOnce({
      releases: [{ version: '1.17' }, { version: '1.14' }],
    });
```
but also I got `docker pull renovate/go:latest` I think there is a problem with this filter in `getDockerTag()`
https://github.com/renovatebot/renovate/blob/c4fa239bfaf61deef30c7d2f95afd77f5d4a7acf/lib/util/exec/docker/index.ts#L107-L109

and if the return from the mocked getPkgReleases version contains `major.minor.patch` like this : 
```
    datasource.getPkgReleases.mockResolvedValueOnce({
      releases: [{ version: '1.17.0' }, { version: '1.14.0' }],
    });
```
the command will be changed to:
`docker pull renovate/go:1.17.0`

I added a test here https://github.com/renovatebot/renovate/blob/c4fa239bfaf61deef30c7d2f95afd77f5d4a7acf/lib/util/exec/docker/index.spec.ts#L102
```
    it('check go version', async () => {
      const releases = [
        { version: '12.0' },
        { version: '13.0' },
        { version: '14.0' },
        { version: '15.0' },
      ];
      getPkgReleases.mockResolvedValueOnce({ releases } as never);
      expect(await getDockerTag('foo', '<14', 'npm')).toBe('12.0');
    });
  });
```
and the returned value from `getDockerTag()` = `latest`